scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,540 @@
+    fn get_generation_by_timestamp_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= ?
+    ORDER BY time DESC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// Given a timestamp of an operation fetch generation that was operating when this operation was performed.
+    /// If no such generation exists, returns None.
+    /// Propagates errors.
+    pub async fn fetch_generation_by_timestamp(
+        &self,
+        time: &chrono::Duration,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_generation_by_timestamp_query());
+        select_consistency(&self.session, &mut query).await?;
+
+        let result = self.session.query(query, (Timestamp(*time),)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    /// Given a timeuuid of a ti of an operation fetch generation that was operating when this operation was performed.
+    /// If no such generation exists, returns None.
+    /// Propagates errors.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let (secs, nanos) = time
+            .to_timestamp()
+            .ok_or("Uuid is not a timeuuid")?
+            .to_unix();
+        let timestamp =
+            chrono::Duration::seconds(secs as i64) + chrono::Duration::nanoseconds(nanos as i64);
+
+        self.fetch_generation_by_timestamp(&timestamp).await
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time ASC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// Given a generation returns the next generation.
+    /// If given generation is currently operating, returns None.
+    /// Propagates errors.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &Generation,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_next_generation_query());
+        select_consistency(&self.session, &mut query).await?;
+
+        let result = self
+            .session
+            .query(query, (Timestamp(generation.0),))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            r#"
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    /// Given a generation return identifiers of all streams of this generation.
+    pub async fn fetch_stream_ids(
```

Comment:
This function and `fetch_all_generations` are almost identical - they do the same thing: prepare a query, execute it, iterate over selected rows and save some value to the vector, return the vector. One could try to abstract this behaviour away but it's probably impossible since required `FromRow` trait is private (not to mention the fact that this may be an overkill for just the two calls). For aesthetical reasons you could at least try to keep the implementations as similar as possible. It's mostly about being consistent in where you put the empty vector initialization - before all the query business or after it.